py3status/modules/sysdata.py

```diff
@@ -196,34 +192,12 @@ def _calc_mem_info(self, unit='GiB', memi=dict, keys=list):
 
         total_mem_kib = memi[keys[0]]
         mem_free = sum([memi[item] for item in keys[1:]])
-
-        try:
-            used_mem_kib = (total_mem_kib - mem_free)
-            used_mem_p = 100 * used_mem_kib / total_mem_kib
-            multiplier = {
-                'KiB': ONE_KIB / ONE_KIB,
-                'MiB': ONE_KIB / ONE_MIB,
-                'GiB': ONE_KIB / ONE_GIB,
-            }
-            if unit.lower() == 'dynamic':
-                # If less than 1 GiB, use MiB
-                if (multiplier['GiB'] * total_mem_kib) < 1:
-                    unit = 'MiB'
-                else:
-                    unit = 'GiB'
-            if unit in multiplier.keys():
-                total_mem = multiplier[unit] * total_mem_kib
-                used_mem = multiplier[unit] * used_mem_kib
-            else:
-                raise ValueError(
-                    'unit [{0}] must be one of: KiB, MiB, GiB, dynamic.'.format(unit))
-        except:
-            total_mem, used_mem, used_mem_p = [float('nan') for i in range(3)]
-            unit = 'UNKNOWN'
-
-        # If total memory is <1GB, results are in megabytes.
-        # Otherwise, results are in gigabytes.
-        return total_mem, used_mem, used_mem_p, unit
+        used_mem_kib = (total_mem_kib - mem_free)
+        used_mem_p = 100 * used_mem_kib / total_mem_kib
+        unit = 'B' if unit == 'dynamic' else unit
+        (total_mem, unit) = self.py3.format_units(total_mem_kib * 1024, unit=unit)
+        (used_mem, _) = self.py3.format_units(used_mem_kib * 1024, unit=unit)
```

Comment:
* We can set units once in `post_config_hook` instead of keep checking here.
* `_` can be useful. Offer `{mem_used_unit}` and/or `{mem_total_unit}`.
* Is it better to add `optimal=0` too for more accuracy?

Reply:
The idea behind the current implementation is: When `dynamic` is used, the first `py3.format_units` chooses the unit to use `MiB`/`GiB` and the second `py3.format_units` uses the same unit (for consistent display).

Going for `{mem_used_unit}` and `{mem_total_unit}` would be the precise solution, but would require deprecating/dropping `{mem_unit}`.

I don't know whether we should add `optimal=0`.